Added some more tests

codeinred opened the pull request:
I added some more tests to the tests/ directory! There are two new tests: 

- A read-write test, where data is read, and then written - this should fail if reading blocks
- A ping-pong test. Two coroutines send messages back and forth, each time reading the response of the other. I think it has some fun output!

Both tests pass.
![image](https://user-images.githubusercontent.com/36810712/134763526-90820730-4baf-417e-99ab-d7dbaa000d53.png)

![image](https://user-images.githubusercontent.com/36810712/134763553-a7a334e3-d134-46cb-beb3-8c34b720b991.png)

CarterLi commented on tests/ping_pong.cpp:
```diff
@@ -0,0 +1,72 @@
+        co_await service.fsync(write_fd, 0);
+
+        int count =
+            co_await service.read(read_fd, buffer.data(), buffer.size(), 0);
```
io_uring functions don't set the global errno, they return -errno on error. See panic_on_error

CarterLi commented on tests/read_write_pipe.cpp:
```diff
@@ -0,0 +1,50 @@
+auto write_to_fd(uio::io_service& service, int fd) -> uio::task<> {
+    std::string_view message = "Hello, world!";
+    fmt::print("Sending data...\n");
+    co_await service.write(fd, message.data(), message.size(), 0)
```
These two tests seems to be similar. You may try something different

CarterLi commented on tests/ping_pong.cpp:
```diff
@@ -0,0 +1,72 @@
+        int count =
+            co_await service.read(read_fd, buffer.data(), buffer.size(), 0);
+        if (count < 0)
+            uio::panic("Bad read", errno);
```
io_uring IO functions don't set the global errno, they return -errno on error. See panic_on_error

CarterLi commented on tests/ping_pong.cpp:
```diff
@@ -0,0 +1,72 @@
+
+#include <liburing/io_service.hpp>
+#include <string_view>
+auto ping(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
```
A newline is required after `#include`s

CarterLi commented on tests/ping_pong.cpp:
```diff
@@ -0,0 +1,72 @@
+    close(read_fd);
+    close(write_fd);
+}
+int main() {
```
Newline between functions please

CarterLi commented on tests/read_write_pipe.cpp:
```diff
@@ -0,0 +1,50 @@
+
+    io.run(t1);
+    io.run(t2);
+}
```
ping_pong.cpp and read_write_pipe.cpp test the same functions ( read / write / pipe ), please try something different.

CarterLi commented on tests/ping_pong.cpp:
```diff
@@ -0,0 +1,72 @@
+auto pong(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
+    std::string_view ping = "pong!";
+    std::array<char, 64> buffer;
+    for (int i = 0; i < 20; i++) {
```
As pong is a receiver, you may test for EOF here

codeinred replied:
Just to confirm: the value returned from `service.read()` 0 on EOF?

CarterLi replied:
`<= 0` should be ok

codeinred commented:
Hi Carter,

I updated tests/ping_pong.cpp in accordance with the formatting guidelines, and I also fixed the error handling code.

> ping_pong.cpp and read_write_pipe.cpp test the same functions ( read / write / pipe ), please try something different.

Regarding this statement, I decided to remove *read_write_pipe* since ping_pong handled it more thoroughly and is a more interesting test. Please let me know if there's anything else you need!

CarterLi commented:
Merged, thanks.
